cpp/src/traversal/opg/frontier_expansion.cu

```diff
@@ -0,0 +1,152 @@
+  __device__ unsigned operator()(unsigned& visited, unsigned& output) {
+    //OUTPUT AND VISITED - common bits between output and visited
+    //OUTPUT AND (NOT (OUTPUT AND VISITED))
+    // - remove common bits between output and visited from output
+    return (output & (~( output & visited )));
+  }
+};
+
+template <typename VT>
+struct bitwise_atomic_or {
+  unsigned * output_frontier_;
+  VT * predecessors_;
+
+  bitwise_atomic_or(
+      unsigned * output_frontier,
+      VT * predecessors) :
+    output_frontier_(output_frontier),
+    predecessors_(predecessors) {}
+
+  __device__ void operator()(VT src, VT dst) {
+    unsigned active_bit = static_cast<unsigned>(1)<<(dst % BitsPWrd<unsigned>);
+    unsigned prev_word =
+      atomicOr(output_frontier_ + (dst/BitsPWrd<unsigned>), active_bit);
+    //If this thread activates the frontier bitmap for a destination
+    //then the source is the predecessor of that destination
+    if (prev_word & active_bit == 0) {
+      predecessors_[dst] = src;
+    }
+  }
+};
+
+struct is_not_equal {
+  unsigned cmp_;
+  unsigned * flag_;
+  is_not_equal(unsigned cmp, unsigned * flag) : cmp_(cmp), flag_(flag) {}
+  __device__ void operator()(unsigned& val) {
+    if (val != cmp_) {
+      *flag_ = 1;
+    }
+  }
+};
+
+template <typename VT, typename ET, typename WT>
+void bfs(raft::handle_t const &handle,
+    cugraph::experimental::GraphCSRView<VT, ET, WT>& graph,
+    VT *predecessors,
+    const VT start_vertex) {
+
+  //We need to keep track if a vertex is visited or its status
+  //This needs to be done for all the vertices in the global graph
+  size_t word_count = number_of_words(graph.number_of_vertices);
+  rmm::device_vector<unsigned> input_frontier(word_count);
+  rmm::device_vector<unsigned> output_frontier(word_count);
+  rmm::device_vector<unsigned> visited(word_count);
+
+  rmm::device_vector<unsigned> frontier_not_empty(1);
+
+  //Load balancer for calls to bfs functors
+  LoadBalanceExecution<VT, ET, WT> lb(handle, graph);
+
+  //BFS Functor for frontier calculation
+  bitwise_atomic_or<VT> bfs_op(output_frontier.data().get(), predecessors);
+
+  //Functor to check if frontier is empty
+  is_not_equal neq(static_cast<unsigned>(0), frontier_not_empty.data().get());
+
+  cudaStream_t stream = handle.get_stream();
+
+  //BFS communications wrapper
+  BFSCommunicator<VT, ET, WT> bfs_comm(handle, word_count);
+
+  //0. 'Insert' starting vertex in the input frontier
+  input_frontier[start_vertex/BitsPWrd<unsigned>] =
+    static_cast<unsigned>(1)<<(start_vertex%BitsPWrd<unsigned>);
+
+  do {
+    //1. Mark all input frontier vertices as visited
+    thrust::transform(rmm::exec_policy(stream)->on(stream),
+        input_frontier.begin(), input_frontier.end(),
+        visited.begin(),
+        visited.begin(),
+        bitwise_or());
+
+    //2. Clear out output frontier
+    thrust::fill(
+        output_frontier.begin(),
+        output_frontier.end(),
+        static_cast<unsigned>(0));
```

Comment:
What will be the work complexity of 1 and 2 for V vertices and P GPUs? Is it O(V) (instead of O(V/P))?

Reply:
Since the partitioning is 1d, every GPU needs to keep track of the global frontier which is V/32 in size. The complexity therefore is O(V)